cfgov/v1/documents.py

```diff
@@ -0,0 +1,201 @@
+        self.categories = categories
+        self.authors = authors
+        self.to_date = to_date
+        self.from_date = from_date
+        self.title = title
+        self.archived = archived
+
+    def filter_topics(self, search):
+        return search.filter("terms", tags__slug=self.topics)
+
+    def filter_categories(self, search):
+        return search.filter("terms", categories__name=self.categories)
+
+    def filter_authors(self, search):
+        return search.filter("terms", authors__slug=self.authors)
+
+    def filter_date(self, search):
+        return search.filter(
+            "range", **{
+                'date_published': {
+                    'gte': self.from_date, 'lte': self.to_date}})
+
+    def filter_archived(self, search):
+        return search.filter("terms", is_archived=self.archived)
+
+    def search_title(self, search):
+        return search.query(
+            "match", title={"query": self.title, "operator": "AND"}
+        )
+
+    def order_results(self, search):
+        total_results = search.count()
+        return search.sort('-date_published')[0:total_results]
+
+    def has_dates(self):
+        return self.to_date is not None and self.from_date is not None
+
+    def apply_specific_filters(self, search):
+        return search
+
+    def search(self):
+        search = FilterablePagesDocument.search()
+        if self.prefix != '':
+            search = search.filter('prefix', url=self.prefix)
+        if self.topics not in ([], '', None):
+            search = self.filter_topics(search)
+        if self.categories not in ([], '', None):
+            search = self.filter_categories(search)
+        if self.authors not in ([], '', None):
+            search = self.filter_authors(search)
+        if self.has_dates:
+            search = self.filter_date(search)
+        if self.archived is not None:
+            search = self.filter_archived(search)
+        if self.title not in ([], '', None):
+            search = self.search_title(search)
+
+        search = self.apply_specific_filters(search)
+        results = self.order_results(search)
+        return results.to_queryset()
+
+
+class EventFilterablePagesDocumentSearch(FilterablePagesDocumentSearch):
+
+    def filter_date(self, search):
+        return search.filter(
+            "range", **{
+                'start_dt': {'gte': self.from_date}}).filter(
+                    "range", **{'end_dt': {'lte': self.to_date}})
+
+
+class EnforcementActionFilterablePagesDocumentSearch(FilterablePagesDocumentSearch):  # noqa: E501
+
+    def __init__(self,
+                 prefix='/', topics=[], categories=[],
+                 authors=[], to_date=None, from_date=None,
+                 title='', archived=None, statuses=[]):
+        super(EnforcementActionFilterablePagesDocumentSearch, self).__init__(
+            prefix, topics, categories, authors,
+            to_date, from_date, title, archived
+        )
+        self.statuses = statuses
```

Comment:
There's a lot of duplication here. Could this method be made more generic, e.g.

```py
def __init__(self, **kwargs):
    self.statuses = kwargs.pop('statuses')
    super().__init__(**kwargs)
```

Reply:
Done!